### PR TITLE
Add linear regression

### PR DESCRIPTION
**Pull request by vaskers5**

### CODE REVIEW

File: lib/models/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from .linear_regression import LinearRegression
```
Comment (ARQtty) [inline lib/models/__init__.py:1]: newline

File: lib/models/linear_regression.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
+from typing import List, Callable
+import numpy as np
+import yaml
+import pandas as pd
+
+
+class LinearRegression:
+    def __init__(self):
+        self.bias = 0
+        self.weights = np.array([])
```
Comment (ARQtty) [inline lib/models/linear_regression.py:10]: зачем пустой?

File: lib/models/linear_regression.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        self.lr = 0
+        self.losses = []
```
Comment (ARQtty) [inline lib/models/linear_regression.py:12]: и все эти переменные public. Зачем?

File: lib/models/linear_regression.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        
+    def train(self, features: np.ndarray,
```
Comment (ARQtty) [inline lib/models/linear_regression.py:14]: все аргументы нужно писать либо в одной строке, либо в одном столбце

File: lib/models/linear_regression.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+              labels: np.ndarray,
+              bias: float = 0.1,
+              lr: float = 0.001,
+              max_epochs: int = 100) -> 'LinearRegression':
+
+        self.bias, self.lr = bias, lr
+        self.weights = np.random.rand(features.shape[1])
```
Comment (ARQtty) [inline lib/models/linear_regression.py:21]: рандом с незафиксированным сидом? Это невоспроизводимый код

File: lib/models/linear_regression.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        self.weights[0] = bias
+        
+        for i in range(max_epochs):
+            old_weights = self.weights.copy()
```
Comment (ARQtty) [inline lib/models/linear_regression.py:25]: здесь создаётся переменная, которая может в итоге не быть использована. Зачем?

File: lib/models/linear_regression.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            predictions = self.predict(features)
+            loss = self.mse_loss(predictions, labels)
+            if loss == float('inf'):
+                self.weights = old_weights
+                break
+            self.losses += [loss]
```
Comment (ARQtty) [inline lib/models/linear_regression.py:31]: .append

File: lib/models/linear_regression.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            grad = -(labels - predictions).dot(features) / (2.0 * features.shape[0])
+            self.weights -= self.lr * grad
+        
+        return self
+
+    def load_weights(self, path_to_weights: str) -> None:
```
Comment (ARQtty) [inline lib/models/linear_regression.py:37]: path должен быть типом Path

Comment (ARQtty) [reply lib/models/linear_regression.py:37]: Если ты такой любитель возвращать LinearRegression, то почему здесь None?

File: lib/models/linear_regression.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        with open(path_to_weights, 'r') as f:
+            last_version = yaml.load(f)
```
Comment (ARQtty) [inline lib/models/linear_regression.py:39]: веса в yaml это точно ок?

File: lib/models/linear_regression.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        self.weights = np.array(last_version['weights'])
+        self.lr = last_version['lr']
+
+    def save_weights(self, path_to_weights: str) -> None:
```
Comment (ARQtty) [inline lib/models/linear_regression.py:43]: path

File: lib/models/linear_regression.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        last_version = {"weights": self.weights.tolist(),
+                        "lr": self.lr}
+        with open(path_to_weights, 'w') as f:
+            yaml.dump(last_version, f)
+            
+    def predict(self, features: np.ndarray) -> np.ndarray:
+        return np.dot(features, self.weights)
```
Comment (ARQtty) [inline lib/models/linear_regression.py:50]: всегда ли у фичей есть дополнительная колонка единиц? Кажется, код может крашиться, если не делать дополнительную предобработку данных

File: lib/models/linear_regression.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    
+    @staticmethod
+    def mse_loss(predict: np.array, labels: np.array) -> float:
```
Comment (ARQtty) [inline lib/models/linear_regression.py:53]: ndarray

File: lib/models/__init__.py (hunk 1 above)
Comment (vaskers5) [reply lib/models/__init__.py:1]: fixed

File: lib/models/linear_regression.py (hunk 4 above)
Comment (vaskers5) [reply lib/models/linear_regression.py:14]: fixed

File: lib/models/linear_regression.py (hunk 9 above)
Comment (vaskers5) [reply lib/models/linear_regression.py:39]: fixed

File: lib/models/linear_regression.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
+from typing import List
+import numpy as np
+from pathlib import Path
+import pickle
+
+np.random.seed(0)
```
Comment (ARQtty) [inline lib/models/linear_regression.py:6]: хорошей ли идеей является создания сида в самом модуле? Получается сид будет сажаться при импорте модуля. А значит два разных запуска модели будут по-разному инициалированы. Получается, этот сид бесполезен

File: lib/models/linear_regression.py (hunk 4 above)
Comment (ARQtty) [reply lib/models/linear_regression.py:16]: image
Нет

File: lib/models/linear_regression.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    def save_weights(self, path_to_weights: Path) -> None:
+        last_version = {"weights": self._weights.tolist(),
+                        "lr": self._lr}
+        with open(path_to_weights, 'wb') as f:
+            pickle.dump(last_version, f)
+            
+    def _predict(self, features: np.ndarray) -> np.ndarray:
+        return np.dot(features, self._weights)
+    
+    def get_loss(self) -> List[float]:
```
Comment (ARQtty) [inline lib/models/linear_regression.py:55]: get_loss подразумевает одно число - loss. Ты возвращаешь множество лоссов - losses. Путаница

File: lib/models/linear_regression.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+    @staticmethod
+    def mse_loss(predict: np.ndarray,
+                 labels: np.ndarray) -> float:
+        dif = np.power(predict - labels, 2)
+        return dif.sum()/(2 * len(predict))
+
+    @staticmethod
+    def _process_data(features: np.ndarray):
+        return np.c_[features, np.ones(features.shape[0])]
```
Comment (ARQtty) [inline lib/models/linear_regression.py:71]: newline

**Review by ARQtty — CHANGES_REQUESTED** [review]
зачем файл .gitattributes?

**Comment by ARQtty** [comment]
хеллоу?

File: lib/models/linear_regression.py (hunk 15 above)
Comment (vaskers5) [reply lib/models/linear_regression.py:71]: fixed

File: lib/models/linear_regression.py (hunk 14 above)
Comment (vaskers5) [reply lib/models/linear_regression.py:55]: fixed

File: lib/models/linear_regression.py (hunk 13 above)
Comment (vaskers5) [reply lib/models/linear_regression.py:6]: fixed

**Review by ARQtty — APPROVED** [review]
great
